proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -3026,19 +3021,758 @@ lemma reply_unlink_ccorres:
+                         in ccorres_cond_both'[where Q'=\<top>])
+              apply (fastforce simp: option_to_0_def valid_reply'_def split: option.splits)
+             apply (rule updateReply_ccorres_lemma3[where P=\<top> and P'=\<top>])
+              apply vcg
+             apply (frule (1) obj_at_cslift_reply)
+             apply (fastforce intro!: rf_sr_reply_update2
+                                simp: typ_heap_simps' creply_relation_def isHead_def)
+            apply (rule ccorres_return_Skip)
+           apply ceqv
+          apply (clarsimp simp: cleanReply_def bind_assoc)
+          apply (rule ccorres_rhs_assoc2)
+          apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+              apply (rule updateReply_ccorres_lemma2[where P=\<top>])
+               apply vcg
+              apply (frule (1) obj_at_cslift_reply)
+              apply (fastforce intro!: rf_sr_reply_update2
+                                 simp: typ_heap_simps' creply_relation_def split: if_splits)
+             apply ceqv
+            apply (rule ccorres_rhs_assoc2)
+            apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+                apply (rule updateReply_ccorres_lemma2[where P=\<top>])
+                 apply vcg
+                apply (frule (1) obj_at_cslift_reply)
+                apply (fastforce intro!: rf_sr_reply_update2
+                                   simp: typ_heap_simps' creply_relation_def isHead_def)
+               apply ceqv
+              apply (ctac (no_vcg) add: reply_unlink_ccorres)
+             apply (wpsimp wp: updateReply_valid_objs')
+            apply (clarsimp simp: guard_is_UNIV_def)
+           apply (rule_tac Q'="\<lambda>_ s. reply_at' replyPtr s \<and> valid_objs' s \<and> no_0_obj' s
+                                     \<and> pspace_aligned' s \<and> pspace_distinct' s
+                                     \<and> weak_sch_act_wf (ksSchedulerAction s) s"
+                        in hoare_post_imp)
+            apply (clarsimp simp: valid_reply'_def)
+           apply (wpsimp wp: updateReply_valid_objs')
+          apply (clarsimp simp: guard_is_UNIV_def)
+         apply (rule_tac Q'="\<lambda>_ s. reply_at' replyPtr s \<and> valid_objs' s \<and> no_0_obj' s
+                                   \<and> pspace_aligned' s \<and> pspace_distinct' s
+                                   \<and> weak_sch_act_wf (ksSchedulerAction s) s"
+                      in hoare_post_imp)
+          apply (clarsimp simp: valid_reply'_def)
+         apply (wpsimp wp: updateReply_valid_objs')
+        apply (clarsimp simp: guard_is_UNIV_def)
+       apply (rule_tac Q'="\<lambda>_ s. valid_reply' reply s \<and> reply_at' replyPtr s \<and> valid_objs' s
+                                 \<and> no_0_obj' s \<and> pspace_aligned' s \<and> pspace_distinct' s
+                                 \<and> weak_sch_act_wf (ksSchedulerAction s) s"
+                    in hoare_post_imp)
```

Comment:
If this pattern of instantiating can't be avoided in a nice way, one proposal is to have it bound to something (would require an Isar wrapper on the lemma) so we see ?something_post three times instead of the whole block.